### PR TITLE
Exclude “default” build-args from image history 

### PR DESCRIPTION
**Pull request by dave-tucker**
**- What I did**

Based on the discussion in https://github.com/docker/docker/pull/30588#issuecomment-283106005. This PR removes any of the "default" build args from the image history.
This is due to these being considered transparent - i.e the result of an image shouldn't differ based on the `HTTP_PROXY` variable.

**- How I did it**

Edited the builder code to make sure that the any `default` variables are filtered from the `saveCmd` **unless there is a corresponding ARG statement**`

**- How to verify it**

Use the ~force~ tests Luke...

Or alternatively.

Dockerfile:
```
FROM busybox
ARG http_proxy
RUN echo "Test build arg filterage"
```
Then

```
$ docker build -t ba-test --build-arg="HTTP_PROXY=http://user:password@proxy.example.com" --build-arg="http_proxy=http://someproxy.example.com" .
$ docker image history ba-test
```

The `HTTP_PROXY` variable will be absent from the history but `http_proxy` will be present.

**- Description for the changelog**

The values of default build time arguments (e.g `HTTP_PROXY`)  are no longer displayed in `docker image history` unless a corresponding `ARG` instruction is written in the Dockerfile.

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by thaJeztah** [comment]
Thanks! ping @tonistiigi ptal; IIRC you mentioned that the build arg should be included if it's explicitly defined in the Dockerfile, but perhaps I misunderstood

**Comment by tonistiigi** [comment]
Yes, I think adding only explicit ones is slightly better than skipping the builtin ones.

**Comment by dave-tucker** [comment]
@thaJeztah @tonistiigi updated to only remove from history if **is** a builtin arg **and** a corresponding `ARG` instruction **is not** present in the Dockerfile.

**Comment by duglin** [comment]
If I'm following this correctly, I disagree with this PR. In particular this:
> This is due to these being considered transparent - i.e the result of an image shouldn't differ based on the HTTP_PROXY variable.

A changed HTTP_PROXY **MUST** change the image being built. We have no idea what RUN cmd will actually make use of this variable and in what way. For all we know the RUN cmd will echo the entire list of env vars into a file and save that in the image.  I would hope a changed HTTP_PROXY would cause a new image to be built.

**Comment by cpuguy83** [comment]
@duglin Honestly I think this is fairly surprising behavior. The cache is there to help the user get faster builds. If the user wants to break the cache, they certainly can do so manually, meanwhile it's not possible to not break the build cache if the user doesn't want to break it.

**Comment by thaJeztah** [comment]
We're discussing this again in the maintainers meeting, and we think that busting the cache when switching to a different proxy is really a corner case. If someone _wants_ it in the history, they need to define it in the `Dockerfile`. If they need to explicitly bust the cache, `--no-cache` is probably the best approach.

**Comment by estesp** [comment]
Not to mention that if we started to itemize all the ways environmental variations **could** affect an image (example: switching networking config on the system hosting the daemon and getting a different DNS server without a recent (hours old) update to a CNAME/A record for a site used to download code), we would be able to never satisfy ourselves that cacheing was even a reasonable commitment based on the information we **do** know.

**Comment by alexellis** [comment]
In the meeting we also said that this should cover defaults such as HTTP_PROXY, NO_PROXY, FTP_PROXY and HTTPS_PROXY. Any new build-args in the future should be excluded from this behavior.

**Comment by duglin** [comment]
My comments were not about **allowing** people to bust the cache, it was about making sure that when the cache is used its because we know for sure that its because we have the exact same inputs. Changing env vars available to a RUN are supposed to bust the cache.

**Comment by tiborvass** [comment]
@dave-tucker can you please add a comment where the list of default build args are defined, so that if we ever add more default build args, it is understood that this new behavior is applied to that whole list. And if that is not desired for some newly added args, then there will have to be two different lists: "default" one and the "exclude_from_history" one. Today there is no reason to make them separate.

**Comment by tiborvass** [comment]
Maybe @cyli or @diogomonica could have a look at this too ?

**Comment by duglin** [comment]
To be clear.. if I have:
```
RUN doit.sh
```
and in doit.sh I have:
```
if [[ -n "$HTTP_PROXY" ]; then
   app1
else
   app2
fi
```
Then with this PR, building once with HTTP_PROXY set and once w/o will yield the exact same results because the cache will be used in the 2nd run even though it should not have been.

**Comment by alexellis** [comment]
@duglin we discussed using `--no-cache` for this. As a long-term proxy user - this would mean changing from the office network to hotel WiFi would invalidate all my images' cache if I did a rebuild.

**Comment by duglin** [comment]
I feel like we're talking about different things here. How does `--no-cache` help someone - aside from basically having them start from scratch?

Let's back up... first, in order for HTTP_PROXY to be used as a build-arg it needs to be specified on the `docker build` command line.  This means the user is explicitly aware of it.  Now, I realize that with this PR we're making it implicit but to me a config file property is just a short-cut so that people don't need to type something over and over - it shouldn't change the semantics of how things work.

So, would people really expect (ignoring any config file):
```
docker build --build-arg HTTP_PROXY=myproxy.com .
```
and
```
docker build .
```
to yield the same results if they used my "doit.sh" Dockerfile example above?  I would claim no. 

Asking the user to know how an image was previously built so that they could add `--no-cache` when the issue is related to an env var the Docker itself knows about and has total control over its value in the container seems more than a little odd to me.

**Comment by estesp** [comment]
@duglin Why would anyone use your `doit.sh` example..it is quite remarkably a stupid thing to do that no one has any example of anyone doing :)

This is exactly what people are expecting to get the same results; the months spent arguing over even giving users `http_proxy` was an understanding that for some people to have "normal" internet access, this is a required feature. What has come up now is that many corporations have usernames and passwords embedded in this variable which are now stored in their (possibly public) images on DockerHub as a historical build layer.  That is extremely unnecessary for a feature that is merely giving them access to the external internet and was never designed and should never be used to make build-time decisions.  It's like saying I might decide to do something different in my Dockerfile because the TZ setting is different. The principle of least surprise is trying to work in our favor here. Some people like to come up with far-fetched reasons to surprise people with behavior, but I'm guessing that kind of Dockerfile is not exactly going to be a viral hit :)

**Comment by duglin** [comment]
I'm not suggesting that we can or should try to take all aspects of a container into account. I've used this:
```
RUN date > build.timestamp
```
as an example of something people shouldn't do in their Dockerfiles if they really want "build.timestamp" to be the last time they ran the build because caching will muck with that logic.

And in cases like that, the limitation is easy to understand... Docker doesn't know what's going on in the cmd string so it can't make an educated guess as to whether or not to use the cache (at least for the cmd string part of it).

This case is different, as I mentioned above. This is about 1) an env var that the user is explicitly mentioning and 2) an env var the Docker is setting in the container, which means it knows whether its been changed.

I've heard people complain about how we have 2 types of env vars during builds:
1 - known, saved in the image, part of cache check
2 - known, not saved in the image, part of cache check
I'm sure they'll be thrilled to know that now we're adding a third:
3 - known, not saved in the image and not part of the cache checking
:-)

**Comment by thaJeztah** [comment]
The proposal is that if the `HTTP_PROXY` is an essential part of the image (i.e., the `doit.sh` example), that people should define it in the image;

```Dockerfile
ARG HTTP_PROXY
COPY doit.sh /
RUN /doit.sh
```

If it's defined in the Dockerfile, it's an explicit choice of the image-author, and is not ignored in the cache.

There probably are edge cases where a different proxy _should_ produce a different result, but my expectation is that 99.9% of the cases, a proxy is just needed to give people internet access. Producing a different image because a proxy was used for those case, would be the same as invalidating the cache if the IP-address changes (/different network), or the DNS server changed (which would far more likely being a differentiator when building)

**Comment by dave-tucker** [comment]
@tiborvass the list of build args used comes from https://github.com/docker/docker/blob/master/builder/dockerfile/builder.go#L40. This is the only place these are defined in the code so this approach should be still be ok if new args are added.

**Comment by thaJeztah** [comment]
@dave-tucker we were discussing, and thought it would be good to put a comment there so that in case new "default" variables are added in future, we should review if those should be stripped from history as well, or if we should "split" the list at that point

**Comment by dave-tucker** [comment]
@thaJeztah @tiborvass added some comments to the code!

**Comment by tiborvass** [comment]
LGTM

**Comment by tiborvass** [comment]
Ping @tonistiigi

**Comment by tonistiigi** [comment]
LGTM

**Comment by tiborvass** [comment]
@thaJeztah how should we proceed with docs for this one? Out of band?

**Comment by thaJeztah** [comment]
I think we need to add some information to https://github.com/docker/docker/blob/master/docs/reference/builder.md#arg, which already describes the default accepted build-args; we should clarify under which conditions the build args are persisted in the image history, and when not.

I see the command-line reference refers to that section; https://github.com/docker/docker/blob/master/docs/reference/commandline/build.md#set-build-time-variables---build-arg so possibly it's not needed there

(orthogonal; I do notice that we don't explicitly warn against using build-args for sensitive data there, which we should)

**Comment by dave-tucker** [comment]
@thaJeztah added some docs 📖

File: docs/reference/builder.md
Comment (thaJeztah) [inline docs/reference/builder.md]: nit: s/will result/results/

File: docs/reference/builder.md
Comment (thaJeztah) [inline docs/reference/builder.md]: Since we're not refering to specific lines from the text, can you;

- remove the line numbers
- add a `Dockerfile` syntax-hint;

        ```Dockerfile
        FROM ubuntu
        ARG HTTP_PROXY

File: docs/reference/builder.md
Comment (thaJeztah) [inline docs/reference/builder.md]: Same here (line numbers, code-hint)

File: docs/reference/builder.md
Comment (thaJeztah) [inline docs/reference/builder.md]: Same here (line numbers, code-hint)

**Review by thaJeztah — CHANGES_REQUESTED** [review]
Thanks! I left some suggestions, but looks like we've got things covered. Possibly @mstanleyjones would have suggestions for wording as well 😃

File: docs/reference/builder.md
Comment (alexellis) [inline docs/reference/builder.md]: Is it worth mentioning `ENV HTTP_PROXY http://.....` would also be cached?

File: docs/reference/builder.md
Comment (alexellis) [inline docs/reference/builder.md]: Should behaviour be the US version (behavior)?

**Comment by dave-tucker** [comment]
Addressed @thaJeztah's comments.
@alexellis' comment re: `behaviour/behavior` still stands. IMO there is only one correct way 🇬🇧 :trollface:

File: docs/reference/builder.md
Comment (mstanleyjones) [inline docs/reference/builder.md]: Is this actually un-knowable? Things like "may be" inject uncertainty "Will they be visible? Won't they? How to know?" Can we explain when they would or wouldn't? Maybe we can just link to the section below? And instead of "may be" change it to "are visible if you specify them as `ARG` lines in the Dockerfile?

File: docs/reference/builder.md
Comment (mstanleyjones) [inline docs/reference/builder.md]: Please use US spelling of "behavior". :) Also prefer active voice "To override this behavior, provide an explicit...."

File: docs/reference/builder.md
Comment (mstanleyjones) [inline docs/reference/builder.md]: Isn't this the same content as above? Why not have the content one place and then provide a link at the other place?

**Review by mdlinville — CHANGES_REQUESTED** [review]

File: docs/reference/builder.md
Comment (thaJeztah) [inline docs/reference/builder.md]: argh, missed this one; s/exluded/excluded/

**Review by thaJeztah — APPROVED** [review]
LGTM

**Comment by dave-tucker** [comment]
Updated docs to address comments from @mstanleyjones and @thaJeztah

**Review by mdlinville — APPROVED** [review]

File: docs/reference/builder.md
Comment (thaJeztah) [inline docs/reference/builder.md]: Would something like this work?

    By default, these pre-defined variables are excluded from the output of
    `docker history`. Excluding them reduces the risk of accidentally leaking
    sensitive authentication information in an `HTTP_PROXY` variable.

    For example, consider building the following Dockerfile using
    `--build-arg HTTP_PROXY=http://user:pass@proxy.lon.example.com`

    ``` Dockerfile
    FROM ubuntu
    RUN echo "Hello World"
    ```

    In this case, the value of the `HTTP_PROXY` variable is not available in the
    `docker history` and is not cached. If you were to change location, and your
    proxy server changed to `http://user:pass@proxy.sfo.example.com`, a subsequent
    build does not result in a cache miss.

    If you need to override this behaviour then you may do so by adding an `ARG`
    statement in the Dockerfile as follows;

    ``` Dockerfile
    FROM ubuntu
    ARG HTTP_PROXY
    RUN echo "Hello World"
    ```

    When building this Dockerfile, the `HTTP_PROXY` is preserved in the
    `docker history`, and changing its value invalidates the build cache.

**Review by thaJeztah — COMMENTED** [review]
Have some suggestions, but I'm ok with addressing them in a follow up 😅

File: docs/reference/builder.md
Comment (mstanleyjones) [inline docs/reference/builder.md]: +1 this sounds much better.

**Comment by dave-tucker** [comment]
@mstanleyjones @thaJeztah thanks for the suggestions! Updated once again...

**Review by thaJeztah — APPROVED** [review]
LGTM, thanks!

**Pull request merged**
